### PR TITLE
Automatically run `bazel mod tidy` in `@rules_go//go`

### LINKED ISSUE #3900: `@rules_go//go get` should trigger `bazel mod tidy`
Bazel 7.1.0 adds a new `bazel mod tidy` that collects and applies `buildozer` fixups and updates the lock file. As [suggested](https://docs.google.com/document/d/13LbK_1WhA4la0eH7yISjnMvXs2cKFXD-adKPu0i0RK0/edit?disco=AAAA2ThnB8U) in the proposal doc, it will be a nice quality of like improvement if `rules_go` trigger `bazel mod tidy` on dependency updates:
> rulesets can provide an improved experience by wrapping the language-native mechanism for adding a dependency and automatically running bazel mod tidy. In the case of rules_go, this simplifies the workflow further to add the dependency via bazel run @rules_go//go get example.com, a wrapper around go get, which automatically runs bazel mod tidy.

### PR DESCRIPTION
**Pull request by fmeum**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

`bazel mod tidy` is run if supported and if relevant files change due to the command.

**Which issues(s) does this PR fix?**

Fixes #3900
Fixes https://github.com/bazelbuild/bazel-gazelle/issues/1749

**Other notes for review**

### CODE REVIEW

**Comment by stefanpenner** [comment]
Just checking, will this also work for users of go.work workspaces?

**Comment by fmeum** [comment]
> Just checking, will this also work for users of go.work workspaces?

This requires adding the relevant files to https://github.com/bazelbuild/bazel-gazelle/blob/7c84c339075a992ea1a6f720899d91c0e047efe2/internal/bzlmod/go_deps.bzl#L657. Maybe that already happens with the fake `go.mod` files we create for a workspace?

File: go/tools/go_bin_runner/main.go [hunk 1]
```diff
@@ -34,8 +42,35 @@ func main() {
...
+	hashesAfter, err := hashWorkspaceRelativeFiles(cfg.DepsFiles)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	diff := diffMaps(hashesBefore, hashesAfter)
+	if len(diff) > 0 {
+		if HasBazelModTidy == "True" {
+			_, _ = fmt.Fprintf(os.Stderr, "\nrules_go: Running 'bazel mod tidy' since %s changed...\n", strings.Join(diff, ", "))
+			err = runProcess([]string{"bazel", "mod", "tidy"}, os.Environ(), cwd)
```
Comment (sluongng) [inline go/tools/go_bin_runner/main.go:66]: FYI: not everybody use the `bazel` binary from $PATH.

Some folks prefer using `bazelisk`, `b`, `bb`, `blaze`.

Could we make this configurable somehow? Or better, could we expose a starlark API for this?

**Review by linzhp — APPROVED** [review]

File: go/tools/go_bin_runner/main.go (hunk 1 above)
Comment (fmeum) [reply go/tools/go_bin_runner/main.go]: Good point, I'm now honoring the `BAZEL` environment variable.

**Pull request merged**